api/file.go

```
@@ -559,14 +560,40 @@ func getPublicLink(c *Context, w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	w.Write([]byte(model.StringToJson(generatePublicLink(c.GetSiteURL(), info))))
+	var publicLink string
+	if utils.Cfg.FileSettings.DriverName == model.IMAGE_DRIVER_S3 && utils.Cfg.FileSettings.UseAwsPresignedUrls {
+		publicLink = generateAwsPresignedUrl(info)
+	} else {
+		publicLink = generatePublicLink(c.GetSiteURL(), info)
+	}
+
+	w.Write([]byte(model.StringToJson(publicLink)))
 }
 
 func generatePublicLink(siteURL string, info *model.FileInfo) string {
 	hash := generatePublicLinkHash(info.Id, *utils.Cfg.FileSettings.PublicLinkSalt)
 	return fmt.Sprintf("%s%s/public/files/%v/get?h=%s", siteURL, model.API_URL_SUFFIX, info.Id, hash)
 }
 
+func generateAwsPresignedUrl(info *model.FileInfo) string {
+	endpoint := utils.Cfg.FileSettings.AmazonS3Endpoint
+	accessKey := utils.Cfg.FileSettings.AmazonS3AccessKeyId
+	secretKey := utils.Cfg.FileSettings.AmazonS3SecretAccessKey
+	secure := *utils.Cfg.FileSettings.AmazonS3SSL
+	bucket := utils.Cfg.FileSettings.AmazonS3Bucket
+	path := info.Path
+	expiry := time.Second * time.Duration(utils.Cfg.FileSettings.AwsPresignedUrlTtl)
+
+	s3Clnt, err := s3.New(endpoint, accessKey, secretKey, secure)
+
+	presignedURL, err := s3Clnt.PresignedGetObject(bucket, path, expiry, nil)
+
+	if err != nil {
```

This looks like an aborted attempt at error handling.  The function should return an error that is checked in the calling function, and used to write an error response.